lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -265,6 +266,88 @@ def add_3ds_exemption(xml, options)
+          'shopperAccountChangeIndicator' => shopper_account_risk_data[:shopper_account_change_indicator],
+          'shopperAccountPasswordChangeIndicator' => shopper_account_risk_data[:shopper_account_password_change_indicator],
+          'shopperAccountShippingAddressUsageIndicator' => shopper_account_risk_data[:shopper_account_shipping_address_usage_indicator],
+          'shopperAccountPaymentAccountIndicator' => shopper_account_risk_data[:shopper_account_payment_account_indicator]
+         ) do
+          add_date_element(xml, 'shopperAccountCreationDate', shopper_account_risk_data.dig(:shopper_account_creation_date, :date))
+          add_date_element(xml, 'shopperAccountModificationDate', shopper_account_risk_data.dig(:shopper_account_modification_date, :date))
+          add_date_element(xml, 'shopperAccountPasswordChangeDate', shopper_account_risk_data.dig(:shopper_account_password_change_date, :date))
+          add_date_element(xml, 'shopperAccountShippingAddressFirstUseDate', shopper_account_risk_data.dig(:shopper_account_shipping_address_first_use_date, :date))
+          add_date_element(xml, 'shopperAccountPaymentAccountFirstUseDate', shopper_account_risk_data.dig(:shopper_account_payment_account_first_use_date, :date))
+        end
+      end
+
+      def add_transaction_risk_data(xml, transaction_risk_data)
+        return unless transaction_risk_data
+
+        xml.transactionRiskData(
+          'shippingMethod' => transaction_risk_data[:shipping_method],
+          'deliveryTimeframe' => transaction_risk_data[:delivery_timeframe],
+          'deliveryEmailAddress' => transaction_risk_data[:delivery_email_address],
+          'reorderingPreviousPurchases' => transaction_risk_data[:reordering_previous_purchases],
+          'preOrderPurchase' => transaction_risk_data[:pre_order_purchase],
+          'giftCardCount' => transaction_risk_data[:gift_card_count]
+        ) do
+          xml.transactionRiskDataGiftCardAmount do
+            amount_hash = {
+              'value' => transaction_risk_data.dig(:transaction_risk_data_gift_card_amount, :amount, :value),
+              'currencyCode' => transaction_risk_data.dig(:transaction_risk_data_gift_card_amount, :amount, :currency),
+              'exponent' => transaction_risk_data.dig(:transaction_risk_data_gift_card_amount, :amount, :exponent)
+            }
+            debit_credit_indicator = transaction_risk_data.dig(:transaction_risk_data_gift_card_amount, :amount, :debit_credit_indicator)
+            amount_hash['debitCreditIndicator'] = debit_credit_indicator if debit_credit_indicator
+            xml.amount(amount_hash)
+          end
+          add_date_element(xml, 'transactionRiskDataPreOrderDate', transaction_risk_data.dig(:transaction_risk_data_pre_order_date, :date))
+        end
+      end
+
+      def add_date_element(xml, name, date)
+        xml.tag! name do
```

Comment:
I left this `tag!` since we can't be sure that `name` won't contain characters that give `Builder::XmlMarkup` trouble.